File: proposed-projects/round-trip-experiment/reverse/20260909T132655Z_2a16d72b/20260910T075017Z_244b957c/ported_implementation/src/grammar-graph/print.ts
import { Color, Colorize } from './colorize.js';
import { getParentStackId } from './get-parent-stack-id.js';
import { isRange, isRuleChar, isRuleRef } from './type-guards.js';
import type { GenericSet } from './generic-set.js';
import type { GraphNode } from './graph-node.js';
import type { GraphPointer } from './graph-pointer.js';

export const printGraphPointer = (
  pointer: GraphPointer,
  colorize: Colorize
): string => {
  const col = colorize;
  return col(`*${getParentStackId(pointer, col)}`, Color.RED);
};

const getChar = (charCode: number): string => {
  const char = String.fromCodePoint(charCode);
  if (char === '\n') {
    return '\\n';
  }
  return char;
};

export const printGraphNode = (
  node: GraphNode,
  colorize: Colorize,
  pointers?: GenericSet<GraphPointer, string>,
  showPosition = false
): string => {
  const col = colorize;
  const rule = node.rule;

  const parts: string[] = [];
  if (showPosition) {
    parts.push(
      col('{', Color.BLUE),
      col(node.id, Color.GRAY),
      col('}', Color.BLUE)
    );
  }

  if (isRuleChar(rule)) {
    const rendered = rule.value.map((v) =>
      isRange(v)
        ? v.map((val) => col(String.fromCodePoint(val), Color.YELLOW)).join('')
        : getChar(v as number)
    );
    parts.push(
      col('[', Color.GRAY),
      col(rendered.join(''), Color.YELLOW),
      col(']', Color.GRAY)
    );
  } else if (isRuleRef(rule)) {
    parts.push(
      col('Ref(', Color.GRAY) +
        col(`${rule.value}`, Color.GREEN) +
        col(')', Color.GRAY)
    );
  } else {
    parts.push(col(rule.type, Color.YELLOW));
  }

  if (pointers && pointers.size > 0) {
    for (const pointer of pointers) {
      const pointerParts: string[] = [];
      if (pointer.node === node) {
        pointerParts.push(pointer.print(col));
      }
      if (pointerParts.length > 0) {
        parts.push(col('[', Color.GRAY));
        parts.push(...pointerParts);
        parts.push(col(']', Color.GRAY));
      }
    }
  }

  const pieces = [parts.join('')];
  if (node.next !== undefined && node.next !== null) {
    const renderedNext = node.next.print(pointers, showPosition, col);
    if (renderedNext) {
      pieces.push(renderedNext);
    }
  }
  return pieces.join(col('-> ', Color.GRAY));
};

File: packages/gbnf-experiment/data/20260909T132655Z_2a16d72b/ported_implementation/gbnf/grammar_graph/print.py
from .colorize import Color
from .get_parent_stack_id import get_parent_stack_id
from .type_guards import is_range, is_rule_char, is_rule_ref


def print_graph_pointer(pointer, colorize=None, **_kwargs) -> str:
    col = colorize
    return col(f"*{get_parent_stack_id(pointer, col)}", Color.RED)


def print_graph_node(node, colorize=None, pointers=None, show_position=False) -> str:
    col = colorize
    rule = node.rule

    parts = []
    if show_position:
        parts.extend(
            [
                col("{", Color.BLUE),
                col(node.id, Color.GRAY),
                col("}", Color.BLUE),
            ]
        )

    if is_rule_char(rule):
        rendered = []
        for v in rule.value:
            if is_range(v):
                rendered.append("".join(col(chr(val), Color.YELLOW) for val in v))
            else:
                rendered.append(_get_char(v))
        parts.extend(
            [
                col("[", Color.GRAY),
                col("".join(rendered), Color.YELLOW),
                col("]", Color.GRAY),
            ]
        )
    elif is_rule_ref(rule):
        parts.append(
            col("Ref(", Color.GRAY)
            + col(f"{rule.value}", Color.GREEN)
            + col(")", Color.GRAY)
        )
    else:
        parts.append(col(rule.type, Color.YELLOW))

    if pointers:
        for pointer in pointers:
            pointer_parts = []
            if pointer.node is node:
                pointer_parts.append(pointer.print(colorize=col))
            if pointer_parts:
                parts.append(col("[", Color.GRAY))
                parts.extend(pointer_parts)
                parts.append(col("]", Color.GRAY))

    pieces = ["".join(parts)]
    if node.next is not None:
        rendered_next = node.next.print(
            pointers=pointers, colorize=col, show_position=show_position
        )
        if rendered_next:
            pieces.append(rendered_next)
    return col("-> ", Color.GRAY).join(pieces)


def _get_char(char_code: int) -> str:
    char = chr(char_code)
    if char == "\n":
        return "\\n"
    return char
